File: Habitica/src/main/java/com/habitrpg/android/habitica/interactors/UseCase.java
package com.habitrpg.android.habitica.interactors;

import com.habitrpg.android.habitica.executors.PostExecutionThread;

import io.reactivex.rxjava3.core.Flowable;

public abstract class UseCase<Q extends UseCase.RequestValues, T> {

    private final PostExecutionThread postExecutionThread;

    protected UseCase(PostExecutionThread postExecutionThread) {
        this.postExecutionThread = postExecutionThread;
    }

    protected abstract Flowable<T> buildUseCaseObservable(Q requestValues);

    public Flowable<T> observable(Q requestValues) {
        return this.buildUseCaseObservable(requestValues)
                .subscribeOn(postExecutionThread.getScheduler())
                .observeOn(postExecutionThread.getScheduler());
    }

    public interface RequestValues {

    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/interactors/UseCase.kt
package com.habitrpg.android.habitica.interactors

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

abstract class UseCase<Q : UseCase.RequestValues?, T> {
    protected abstract suspend fun run(requestValues: Q): T
    suspend fun callInteractor(requestValues: Q): T {
        return withContext(Dispatchers.Main) {
            run(requestValues)
        }
    }

    interface RequestValues
}
